dimod/sampleset.py

```diff
@@ -789,6 +789,75 @@ def from_samples_bqm(cls, samples_like, bqm, **kwargs):
 
         return cls.from_samples(samples_like, energy=energies, vartype=bqm.vartype, **kwargs)
 
+    @classmethod
+    def from_samples_cqm(cls, samples_like, cqm, rtol=1e-6, atol=1e-8, **kwargs):
+        """Build a sample set from raw samples and a constrained quadratic model.
+
+        The constrained quadratic model is used to calculate energies and feasibility.
+
+        Args:
+            samples_like:
+                A collection of raw samples. 'samples_like' is an extension of NumPy's array_like.
+                See :func:`.as_samples`.
+
+            cqm (:obj:`.ConstrainedQuadraticModel`):
+                A constrained quadratic model.
+
+            rtol (float, optional, default=1e-6):
+                Relative tolerance for constraint violation.
+                See :meth:`.ConstrainedQuadraticModel.check_feasible` for more information.
+
+            atol (float, optional, default=1e-8):
+                Absolute tolerance for constraint violations.
+                See :meth:`.ConstrainedQuadraticModel.check_feasible` for more information.
+
+            info (dict, optional):
+                Information about the :class:`SampleSet` as a whole formatted as a dict.
+
+            num_occurrences (array_like, optional):
+                Number of occurrences for each sample. If not provided, defaults to a vector of 1s.
+
+            aggregate_samples (bool, optional, default=False):
+                If True, all samples in returned :obj:`.SampleSet` are unique,
+                with `num_occurrences` accounting for any duplicate samples in
+                `samples_like`.
+
+            sort_labels (bool, optional, default=True):
+                Return :attr:`.SampleSet.variables` in sorted order. For mixed
+                (unsortable) types, the given order is maintained.
+
+            **vectors (array_like):
+                Other per-sample data.
+
+        Returns:
+            :obj:`.SampleSet`
+
+        Examples:
+
+            >>> cqm = dimod.ConstrainedQuadraticModel()
+            >>> x, y, z = dimod.Binaries(['x', 'y', 'z'])
+            >>> cqm.set_objective(x*y + 2*y*z)
+            >>> label = cqm.add_constraint(x*y == 1, label='constraint_1')
+            >>> sampleset = dimod.SampleSet.from_samples_cqm({'x': 0, 'y': 1, 'z': 1}, cqm)
+
+        """
+        # more performant to do this once, here rather than again in cqm.objective.energies
+        # and in cls.from_samples
+        samples_like = samples, labels = as_samples(samples_like)
+
+        energies = cqm.objective.energies(samples_like)
+
+        is_satisfied = [[datum.violation <= atol + rtol*abs(datum.rhs_energy)
+                         for datum in cqm.iter_constraint_data((s, labels))] for s in samples]
+        is_feasible = [all(satisfied) for satisfied in is_satisfied]
+
+        info = kwargs.get('info', {})
+        info['constraint_labels'] = list(cqm.constraints.keys())
```

Comment:
this makes me a bit nervous. I know that dicts are ordered in Python 3.7+, and that [.iter_constraint_data() uses that ordering](https://github.com/dwavesystems/dimod/blob/003253f5a28419ed7664042bb7dd3644c78832c0/dimod/constrained.py#L975) so it's safe, but this feels like a place that a bug could easily creep in.

Can we either read the label from `datum`, or document in `iter_constraint_data` that it's guaranteed to use the same order as `.constraints`, or both? Perhaps a test as well.

Reply:
Thanks for that feedback, I was thinking about this also.  Here is a question relating to possible implementations of this.  What is the expected behavior for the case of `samples_like` being empty?

`SampleSet.from_samples` supports empty `samples_like`.  But `SampleSet.from_samples_bqm` does not:

```python
>>> bqm = dimod.BinaryQuadraticModel.from_ising({}, {('a', 'b'): -1})
>>> sampleset = dimod.SampleSet.from_samples_bqm([], bqm)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/timdoyle/src/dimod/dimod/sampleset.py", line 788, in from_samples_bqm
    energies = bqm.energies(samples_like)
  File "/Users/timdoyle/src/dimod/dimod/binary/binary_quadratic_model.py", line 1079, in energies
    return self.data.energies(samples_like, dtype=dtype)
  File "dimod/binary/cybqm/cybqm_template.pyx.pxi", line 378, in dimod.binary.cybqm.cybqm_float64.cyBQM_template.energies
    return np.asarray(self._energies[np.float64_t](samples, labels), dtype=dtype)
  File "dimod/binary/cybqm/cybqm_template.pyx.pxi", line 341, in dimod.binary.cybqm.cybqm_float64.cyBQM_template._energies
    raise ValueError(f"missing variable {self.variables[si]!r} in sample(s)")
ValueError: missing variable 'a' in sample(s)
```

So, the question becomes, should `SampleSet.from_samples_cqm` behave like `SampleSet.from_samples_bqm` (effectively not support empty `samples_like`) or should it behave like `SampleSet.from_samples`?

I am leaning toward the latter because the `ValueError` raised by `SampleSet.from_samples_bqm` seems like an accident rather than intention.  That interpretation raises another question: should we change the behavior of `SampleSet.from_samples_bqm` to support empty `samples_like` also?  I would lean toward "yes" to that also.  In any case, having consistent behavior in handling empty `samples_like` among all three methods seems desirable.

Reply:
Yes, IMO the latter makes sense. I think that's another bug :facepalm:

Reply:
> Can we either read the label from `datum`, or document in `iter_constraint_data` that it's guaranteed to use the same order as `.constraints`, or both? Perhaps a test as well.

✔️ 
Went with the "document that ``iter_constraint_data`` is gauranteed to use the same order" approach and added a unit test to cover that.